File: clojure/src/life.clj
(ns life
  (:require [clojure.string :as str]))

(defn load-env [path]
  (try
    (->> (slurp path)
         str/split-lines
         (map str/trim)
         (remove #(or (str/blank? %) (str/starts-with? % "#")))
         (map #(str/split % #"=" 2))
         (filter #(= 2 (count %)))
         (into {}))
    (catch Exception _
      {})))

(def env (load-env "../.env"))

(defn env-int [key default]
  (try
    (Integer/parseInt (get env key (str default)))
    (catch Exception _ default)))

(defn env-double [key default]
  (try
    (Double/parseDouble (get env key (str default)))
    (catch Exception _ default)))

(def width (env-int "WIDTH" 40))
(def height (env-int "HEIGHT" 20))
(def speed (env-double "SPEED" 0.15))
(def density (env-double "DENSITY" 0.25))

(defn create-grid []
  (vec
    (for [_ (range height)]
      (vec
        (for [_ (range width)]
          (< (rand) density))))))

(defn alive? [grid x y]
  (get-in grid [y x] false))

(defn count-neighbors [grid x y]
  (count
    (for [dy [-1 0 1]
          dx [-1 0 1]
          :when (not= [dx dy] [0 0])
          :let [nx (+ x dx)
                ny (+ y dy)]
          :when (alive? grid nx ny)]
      true)))

(defn next-cell [grid x y]
  (let [alive (alive? grid x y)
        neighbors (count-neighbors grid x y)]
    (or
      (and alive (or (= neighbors 2) (= neighbors 3)))
      (and (not alive) (= neighbors 3)))))

(defn next-generation [grid]
  (vec
    (for [y (range height)]
      (vec
        (for [x (range width)]
          (next-cell grid x y))))))

(defn clear-screen []
  (print "\033[H\033[2J")
  (flush))

(defn render [grid]
  (clear-screen)
  (doseq [row grid]
    (println
      (apply str
             (map #(if % "█" " ") row)))))

(defn -main []
  (loop [grid (create-grid)]
    (render grid)
    (Thread/sleep (long (* speed 1000)))
    (recur (next-generation grid))))

(-main)

File: zig/life.zig
const std = @import("std");
const envmod = @import("env.zig");

const Grid = []bool;

fn isAlive(grid: Grid, width: usize, height: usize, x: isize, y: isize) bool {
    if (x < 0 or y < 0) return false;

    const ux: usize = @intCast(x);
    const uy: usize = @intCast(y);

    if (ux >= width or uy >= height) return false;

    return grid[uy * width + ux];
}

fn countNeighbors(grid: Grid, width: usize, height: usize, x: usize, y: usize) usize {
    var count: usize = 0;

    var dy: isize = -1;
    while (dy <= 1) : (dy += 1) {
        var dx: isize = -1;
        while (dx <= 1) : (dx += 1) {
            if (dx == 0 and dy == 0) continue;

            if (isAlive(grid, width, height,
                @as(isize, @intCast(x)) + dx,
                @as(isize, @intCast(y)) + dy))
                {
                    count += 1;
                }
        }
    }

    return count;
}

fn nextGeneration(current: Grid, next: Grid, width: usize, height: usize) void {
    var y: usize = 0;
    while (y < height) : (y += 1) {
        var x: usize = 0;
        while (x < width) : (x += 1) {
            const alive = current[y * width + x];
            const neighbors = countNeighbors(current, width, height, x, y);

            next[y * width + x] =
            (alive and (neighbors == 2 or neighbors == 3)) or
                (!alive and neighbors == 3);
        }
    }
}

fn render(grid: Grid, width: usize, height: usize) void {
    std.debug.print("\x1b[H\x1b[2J", .{});

    var y: usize = 0;
    while (y < height) : (y += 1) {
        var x: usize = 0;
        while (x < width) : (x += 1) {
            std.debug.print("{s}", .{if (grid[y * width + x]) "█" else " "});
        }
        std.debug.print("\n", .{});
    }
}

pub fn main() !void {
    const allocator = std.heap.page_allocator;

    var env = try envmod.load(allocator, "../.env");
    defer envmod.free(allocator, &env);

    const width = envmod.getInt(env, "WIDTH", 40);
    const height = envmod.getInt(env, "HEIGHT", 20);
    const speed = envmod.getFloat(env, "SPEED", 0.15);
    const density = envmod.getFloat(env, "DENSITY", 0.25);

    var current = try allocator.alloc(bool, width * height);
    defer allocator.free(current);

    var next = try allocator.alloc(bool, width * height);
    defer allocator.free(next);

    var prng = std.Random.DefaultPrng.init(@intCast(std.time.timestamp()));
    const random = prng.random();

    for (current) |*cell| {
        cell.* = random.float(f64) < density;
    }

    while (true) {
        render(current, width, height);
        nextGeneration(current, next, width, height);

        const tmp = current;
        current = next;
        next = tmp;

        std.time.sleep(@intFromFloat(speed * std.time.ns_per_s));
    }
}